sqlccl: allow canceling of IMPORT jobs

madelynnblue opened the pull request:
Special case import jobs in the job cancel function. This allows them
to be marked as canceled, even though they are not resumable. This
is a short term fix (for some unknown definition of short term) until
the jobs API can be yet again improved to support jobs that start jobs.

IMPORT needed to be taught about monitoring the progress callback so
it could cancel itself.

While here, improve the error message around attempting to
pause/cancel/resume schema change jobs.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/20343)
<!-- Reviewable:end -->

dianasaur323 commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> On further thought, it might just make sense to make those job table fixes in conjunction with thinking about how schema changes work with that in the next release. 

---

Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20343#-:-L-ClmmGhqQ1I6kXanVl:b-uostp5)*
<!-- Sent from Reviewable.io -->

mjibson commented:
Good idea. Agree.

benesch commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

![s0105150\_sc7-2.jpeg](https://files.reviewableusercontent.io/images/reviewable/882976/dJUHO-TzeNYKlSVd-JwOkEbH/s0105150_sc7-2.jpeg)

---

Reviewed 1 of 5 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[pkg/ccl/sqlccl/backup_test.go, line 964 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/20343#-L-D_Xta0698INmk4qKs:-L-D_Xta0698INmk4qKt:b7bbh24) ([raw file](https://github.com/cockroachdb/cockroach/blob/76f7f59f838babad68c832d2d396ce60c739123e/pkg/ccl/sqlccl/backup_test.go#L964)):*
> ```Go
> 		// Check that executing again succeeds. This won't work if the first import
> 		// was not successfully canceled.
> 		sqlDB.Exec(t, query, cancelDir)
> ```

This isn't the ideal place for this test, but ¯\_(ツ)_/¯ . The setup for testing this is annoying enough that I'm ok with it. If we find ourselves doing more of this, though, let's extract a jobs_test.go in sqlccl, and stick all the jobs-related backup/restore/import tests there.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/20343#-:-L-DZkr2czXKiMCJwGO_:bpmitxu)*
<!-- Sent from Reviewable.io -->

madelynnblue committed:
sqlccl: allow canceling of IMPORT jobs

Special case import jobs in the job cancel function. This allows them
to be marked as canceled, even though they are not resumable. This
is a short term fix (for some unknown definition of short term) until
the jobs API can be yet again improved to support jobs that start jobs.

IMPORT needed to be taught about monitoring the progress callback so
it could cancel itself.

While here, improve the error message around attempting to
pause/cancel/resume schema change jobs.

Release note (sql change): allow canceling of IMPORT jobs.
